### PR TITLE
Fix new closures code such that mirahc compiles, remove old closures implementation.

### PR DESCRIPTION
**Pull request by felixvf**

### CODE REVIEW

File: src/org/mirah/builtins/object_extensions.mirah [hunk 1]
```diff
@@ -16,7 +16,7 @@
 package org.mirah.builtins
 
 import mirah.lang.ast.*
-import org.mirah.typer.ClosureBuilder
+import org.mirah.typer.BetterClosureBuilder
```
Comment (baroquebobcat) [inline src/org/mirah/builtins/object_extensions.mirah]: looks like this isn't necessary.

File: src/org/mirah/builtins/object_extensions.mirah [hunk 2]
```diff
@@ -216,21 +216,10 @@ class ObjectExtensions
...
 
   macro def lambda(type:TypeName, block:Block)
-  # TODO, just create a SyntheticLambdaDefinition that the ClosureBuilder picks up
-    builder = ClosureBuilder.new(@mirah.typer)
-    scope = @mirah.typer.scoper.getScope(@call)
-    resolved = @mirah.type_system.get(scope, type.typeref).resolve
-    # TODO, do something better than this
-    raise "can't build lambda for #{type} #{resolved}" if resolved.kind_of? ErrorType
-    builder.prepare(block, resolved)
+    SyntheticLambdaDefinition.new(@call.position,type,block)
```
Comment (baroquebobcat) [inline src/org/mirah/builtins/object_extensions.mirah]: Could you add whitespace between the ',' and the following argument? Same below.

File: src/org/mirah/jvm/compiler/constructor_cleanup.mirah [hunk 3]
```diff
@@ -55,11 +55,19 @@ class ConstructorCleanup < SimpleNodeVisitor
...
       child = node.get(i)
-      unless child.kind_of?(ClassDefinition)
-        return child.accept(self, arg)
-      end
+      res = child.accept(self, arg)
+      return res if res
     end
   end
+  
+  def visitClassDefinition(node, arg)
```
Comment (baroquebobcat) [inline src/org/mirah/jvm/compiler/constructor_cleanup.mirah]: Could you add a comment to this and the closure definition one noting that class definitions inside constructors shouldn't be cleaned up?

File: src/org/mirah/jvm/mirrors/better_scope.mirah [hunk 4]
```diff
@@ -601,6 +652,14 @@ class MethodScope < BetterScope
     @locals = Locals.new
     @imports = ImportsAndSearchPackages.new
   end
+  
+  def initialize(source: ClosureScope, context: Node)
+    super(context)
+    @scoper     = source.internal_scoper
+    @locals     = source.internal_locals
+    @imports    = source.internal_imports
```
Comment (baroquebobcat) [inline src/org/mirah/jvm/mirrors/better_scope.mirah]: I think these should defer to the source's, but not reuse them. Otherwise, if you have something like

```
foo do
  def f(a); end
  def g(a); end
end
```
wouldn't that cause an issue?

File: src/org/mirah/jvm/mirrors/better_scope.mirah [hunk 5]
```diff
@@ -79,6 +79,10 @@ class Locals
     end
     type
   end
+  
+  def has_local_type(name:String):boolean
+    @local_types.containsKey(name)
```
Comment (baroquebobcat) [inline src/org/mirah/jvm/mirrors/better_scope.mirah]: This seems like it should be unnecessary. Is this because has_local uses `@declared_locals`, which has elements removed when they resolve as an error?
Maybe we could see if changing `has_local` to check the key set instead would be better.

I'm also not sure why Locals removes elements when they are in error. I find that a little surprising.

File: src/org/mirah/jvm/mirrors/method_lookup.mirah [hunk 6]
```diff
@@ -296,7 +296,7 @@ class MethodLookup
     potentials = gatherMethods(target, name)
     is_static_scope = (
         scope && scope.selfType && target == scope.selfType.resolve)
-    if includeStaticImports && potentials.isEmpty && is_static_scope
+    if includeStaticImports && potentials.isEmpty # && is_static_scope # FIXME: Why require is_static_scope?
```
Comment (baroquebobcat) [inline src/org/mirah/jvm/mirrors/method_lookup.mirah]: I changed this in another patch. I think they were just being overly careful.

File: src/org/mirah/tool/mirah_compiler.mirah [hunk 7]
```diff
@@ -164,7 +164,9 @@ class MirahCompiler implements JvmBackend
 
     sorted_asts.each do |node: Node|
       begin
+        node.accept(AstChecker.new,nil) if AstChecker.enabled
```
Comment (baroquebobcat) [inline src/org/mirah/tool/mirah_compiler.mirah]: whitespace. Also, you could extract this into a static method on AstChecker. eg `AstChecker.maybe_check(node)`

File: src/org/mirah/tool/mirah_compiler.mirah [hunk 8]
```diff
@@ -164,7 +164,9 @@ class MirahCompiler implements JvmBackend
 
     sorted_asts.each do |node: Node|
       begin
+        node.accept(AstChecker.new,nil) if AstChecker.enabled 
         @typer.infer(node, false)
+        node.accept(AstChecker.new,nil) if AstChecker.enabled
```
Comment (baroquebobcat) [inline src/org/mirah/tool/mirah_compiler.mirah]: whitespace

File: src/org/mirah/typer/better_closures.mirah [hunk 9]
```diff
@@ -389,27 +434,35 @@ class BetterClosureBuilder
...
-    # ignore closures with no parents, they aren't in the final AST, maybe
-    closures.removeAll closures_to_skip
-
-    i = 0
-    closures.each do |entry: Entry|
-      @@log.fine "insert_closure #{entry.getKey} #{entry.getValue} #{i}"
-      i += 1
-
-      block = Block(entry.getKey)
+      binding_generated = adjuster.adjust enclosing_b, block
```
Comment (baroquebobcat) [inline src/org/mirah/typer/better_closures.mirah]: `binding_generated` is never used. Could you revert the change to cause adjust to return a bool, and rm it?

File: src/org/mirah/typer/better_closures.mirah [hunk 10]
```diff
@@ -389,27 +434,35 @@ class BetterClosureBuilder
...
+#    closures.removeAll closures_to_skip
+#
+#    i = 0
+#    closures.each do |entry: Entry|
+#      @@log.fine "insert_closure #{entry.getKey} #{entry.getValue} #{i}"
+#      i += 1
+
+      if AstChecker.enabled
+        scripts.each do |s: Script|
+          s.accept(AstChecker.new,nil)
```
Comment (baroquebobcat) [inline src/org/mirah/typer/better_closures.mirah]: whitespace

File: src/org/mirah/typer/better_closures.mirah [hunk 11]
```diff
@@ -477,14 +523,22 @@ class BetterClosureBuilder
 
       
 
-      parent = CallSite(block.parent)
-      replace_block_with_closure_in_call parent, block, new_node
+      if block.parent.kind_of?(CallSite)
+        parent = CallSite(block.parent)
+        replace_block_with_closure_in_call parent, block, new_node
+      else
```
Comment (baroquebobcat) [inline src/org/mirah/typer/better_closures.mirah]: I think I'd prefer if this was a elsif checking the type against SyntheticLambdaDefinition, with an else that raises an error.

File: src/org/mirah/typer/better_closures.mirah [hunk 12]
```diff
@@ -922,7 +1011,11 @@ enclosing_scope = get_scope(enclosing_body)
     return_type = makeSimpleTypeName(block.position, mtype.returnType)
     block_method = MethodDefinition.new(block.position, name, args, return_type, nil, nil)
 
-    block_method.body = NodeList(block.body.clone)
+    closure_scope = ClosureScope(get_inner_scope(block))
+
+    block_body        = block.body
+#   block.body        = nil        # prevent cloning of block_body, such that later calls to findAncestor actually can find a non-nil ancestor
```
Comment (baroquebobcat) [inline src/org/mirah/typer/better_closures.mirah]: If this isn't necessary, could you rm it?

File: src/org/mirah/typer/synthetic_lambda_future.mirah [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+
+package org.mirah.typer
+
+import mirah.lang.ast.SyntheticLambdaDefinition
+import mirah.lang.ast.Position
+
+import org.mirah.jvm.mirrors.MirrorType
+
+class SyntheticLambdaFuture < BaseTypeFuture
+  def initialize(supertype:TypeFuture,block:BlockFuture,position:Position)
```
Comment (baroquebobcat) [inline src/org/mirah/typer/synthetic_lambda_future.mirah]: whitespace between `,`s and the following parameter.

File: src/org/mirah/typer/typer.mirah [hunk 14]
```diff
@@ -574,6 +569,7 @@ class Typer < SimpleNodeVisitor
     end
     enclosing_node = node.findAncestor {|n| n.kind_of?(MethodDefinition) || n.kind_of?(Script)}
     if enclosing_node.kind_of? MethodDefinition
+      return nil if isMethodInBlock(MethodDefinition(enclosing_node)) # return types are not supported currently for methods which act as templates
```
Comment (baroquebobcat) [inline src/org/mirah/typer/typer.mirah]: This is because the method will be re-inferred during closure creation right?

File: src/org/mirah/typer/typer.mirah [hunk 15]
```diff
@@ -1134,38 +1130,51 @@ class Typer < SimpleNodeVisitor
...
+      scope_around_block = scopeOf(block)
+      scope              = addScopeUnder(mdef)
+      scope.selfType     = scope_around_block.selfType
+      scope.parent       = scope_around_block # We may want to access variables available in the scope outside of the block.
+      scope
       infer(mdef.body, false)
-      type.returnType.assign(@types.getVoidType, mdef.position)
-    else
-      type.returnType.assign(infer(mdef.body), mdef.body.position)
+      nil
```
Comment (baroquebobcat) [inline src/org/mirah/typer/typer.mirah]: Maybe note that the method's type will be inferred during closure creation?

File: src/org/mirah/typer/typer.mirah [hunk 16]
```diff
@@ -1134,38 +1130,51 @@ class Typer < SimpleNodeVisitor
...
+      end
+      type
+    else  # We are a method defined in a block. We are just a template for a method in a ClosureDefinition
+      block = Block(mdef.parent.parent)
+      @@log.finest "Method #{mdef} is member of #{block}"
+      scope_around_block = scopeOf(block)
+      scope              = addScopeUnder(mdef)
+      scope.selfType     = scope_around_block.selfType
+      scope.parent       = scope_around_block # We may want to access variables available in the scope outside of the block.
+      scope
```
Comment (baroquebobcat) [inline src/org/mirah/typer/typer.mirah]: rm since it's not doing anything.

File: src/org/mirah/util/ast_checker.mirah [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+package org.mirah.util
+
+import java.util.ArrayDeque
+import mirah.lang.ast.Node
+import mirah.lang.ast.NodeScanner
+
+class AstChecker < NodeScanner
```
Comment (baroquebobcat) [inline src/org/mirah/util/ast_checker.mirah]: Could you add a comment explaining what this checks?

**Comment by baroquebobcat** [comment]
It looks really good. I've got a few nitpicks, but I'm really excited to build on the work you've put together here!

File: src/org/mirah/typer/typer.mirah (hunk 14 above)
Comment (felixvf) [inline src/org/mirah/typer/typer.mirah]: > This is because the method will be re-inferred during closure creation right?

No. This is for handling methods which are part of anonymous classes which are currently (mis-)using block syntax, as explained in https://github.com/mirah/mirah/issues/354 .
That's one reason why I think that block syntax and anonymous classes syntax should be separated.

These "special" MethodDefinitions seem to require running Typer#infer() on them (in order to know that they have free variables which should be part of a binding of an enclosing block), but once they have been copied into the AST of the anonymous class, the Typer will run on these MethodDefinitions' clones again, anyway.

If this line was not there, the next line's `infer(enclosing_node)` would result in an endless recursion.

File: src/org/mirah/jvm/mirrors/better_scope.mirah (hunk 4 above)
Comment (felixvf) [inline src/org/mirah/jvm/mirrors/better_scope.mirah]: No, this is for converting a ClosureScope into a MethodScope when the closure is converted into a method, where the body of the closure becomes the body of the method. Hence, the old ClosureScope will not be reused but just discarded. And would happen only if the block does not contain methods at all.

File: src/org/mirah/typer/typer.mirah (hunk 14 above)
Comment (baroquebobcat) [inline src/org/mirah/typer/typer.mirah]: I think it'll only cause endless recursion for the new code. In the existing version, the method's future is set at https://github.com/mirah/mirah/blob/master/src/org/mirah/typer/typer.mirah#L1153

We could make that more explicit by using `getInferredType` here, and `learnType` there, since neither will cause additional traversal.

Is it necessary to not give nested methods a future, or is it just that they don't need the future yet? Maybe we could still set a MethodFuture for them, but let it get overridden during the re-infer. Ultimately, I think we want to have all the type futures generated at this stage, and just copy them over for the closure building, but we can do that incrementally.

Comment (felixvf) [inline src/org/mirah/typer/typer.mirah]: They do not need the future at all, because the MethodDefinition subtrees are currently just use as a template, and clone()d into new MethodDefinitions, which are then properly infer()ed on.

I've tried to replace the cloning by mere relinking of the AST, but failed in the first attempt due to compiler intricacies. As anonymous classes are currently a not-often-used feature, I think that this optimization can be implemented later.

File: src/org/mirah/jvm/mirrors/better_scope.mirah (hunk 4 above)
Comment (baroquebobcat) [inline src/org/mirah/jvm/mirrors/better_scope.mirah]: Cool. I get it now.
